### ISSUE TITLE
:warning: "Any" columns are subject to data corruption (via Python mutability)!

### ISSUE THREAD

**Issue by combinatorist** [gristlabs/grist-core#875]
# TL;DR
References in a formula column to another (Any type) field actually reference the in-memory python object contained there and can semi-permanently modify that (in volatile memory).

# Steps to reproduce:
1. copy this [public document](https://tim-eccleston.getgrist.com/mZA5uZbpfjfe/mutable-python-bug-before?utm_id=share-doc)
2. modify it according to this [screencast](https://www.dropbox.com/scl/fi/7wv5gj5iwhreae8xge184/Screen-recording-2024-02-27-16.31.46.webm?rlkey=mc2wdpqslsqmz30fc9j3513e2&dl=0)
  1. edit a value in the `Stowaway` column.
  2. notice the new value appended to the `Ship` column (mutable)
  3. "Duplicate" (or save an unsaved copy of) the document
  4. notice that the original value of `Stowaway` disappears from `Ship` (volatile memory)
  5. and your new value remains (reloaded from storage)

# Implications
- very tricky to bug to trace when you accidentally modify a column via another one
- very hard to (ab)use as a feature since it exists only in volatile memory

**Comment by dsagal** [comment]
Thanks @combinatorist. This is a strange beast; not exactly a bug in Grist, arguably a bug in the formula, but admittedly a situation that feels wrong for a user, and hard to understand and deal with. To generalize it, it's about the conflict between the concept of spreadsheet formulas, which are inherently functional in nature (stateless, output depends only on their inputs, free of side-effects), and Python, which is not a functional language.

In Grist formulas, the general rule should always be: avoid side-effects, return a value based only on the inputs.

A formula like `$Ship.append($Stowaway)` is all side-effects, and that's the cause of trouble. Instead, the formula _should be_ `$Ship + [$Stowaway]`, which returns the extended list without changing (mutating) the original.

If Grist used a pure functional language instead of Python, such side effects wouldn't be a problem, but as long as Python is the language, such code as `$Ship.append($Stowaway)` does seem natural, and in fact what one would find online looking for "how to add a value to a list". So the big question is the implication you listed: how to detect this, and guide the user towards correct usage.

One option is to drop `Any` type entirely, and always cast the results of formulas to some Grist-specific type for which Grist can enforce some behavior. But there are some legitimate (even if hacky) cases though when a formula could return a rich object (e.g. a Python function or class instance), so a change like that could break existing documents and prevent interesting use cases.

A smaller change would be to catch common known types, e.g. if a formula value is a `list`, `dict`, or `set`, its use in another formula could get wrapped into some Python wrapper that intercepts calls to mutating methods (like `.append()`) and raises an exception that explains the issue and the alternative.

Interested in your thoughts.

**Comment by combinatorist** [comment]
Hi @dsagal,
Thanks for your thoughtful response. I definitely wouldn't want to get rid of the `Any` data type either.

I guess I was thinking more of a guard rail that would simply use a copy of data. For example, it would apply the `.append` in `Smuggler` to a copy of the data found in `Ship`. But deep copies are expensive and you wouldn't want an advanced user slowed down by the performance losses when it isn't necessary.

**Comment by ricpol** [comment]
Wow, this one's an oddball... :-)
My instinct would be to consider this as a non-issue. I mean, it's a very well-know gotcha moment when dealing with mutable types in Python, like:

```
>>> def foo(x=[]):
        x.append(1)
        return x

>>> foo(); foo(); foo()
```

I dare to say that @combinatorist example would be a code smell for the average hacker, and a _Python_ beginner would never came up with this _in Grist_... but of course, you never know. 
From the point of view of a spreadsheet user, one should have a sense that spreadsheets usually work by "pulling" values from other cells, not the other way around.
So, if one attempts this, I guess they already know they're in dark magic territory, either if they are Python or spreadsheet users. 

Rather, I confess that I was a little puzzled at first by the Grist "any" column type being permissive enough to allow for mutable types... To my eye, the "mistake" in the example is less picking the wrong operation (append instead of add), than putting a list in the Ship field instead of a tuple in the first place. If you use a tuple, you are forced to pick the "right" operation. Better yet, if you try pushing your data and your luck with something like "= $Ship += $Stowaway" you get a clear Syntax Error.

Then again, I can see that checking for mutability can be tricky and expensive. At the end of the day, it doesn't hurt putting a big disclaimer in the documentation, like "please do not use in Grist formulas Python functions that mutate values in place"... even better, "please do not use Python mutables in Grist formulas"... but no further than this. If you want the power of Python in a spreadsheet formula (as you do!), then you get the power of Python in a spreadsheet formula.
